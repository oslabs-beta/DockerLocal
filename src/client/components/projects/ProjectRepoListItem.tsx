import React, { useEffect, useState } from 'react';

import { ProjectRepoListItemProps, Project, Repo } from '../../../types/types';
import { findActiveProject } from '../../helpers/projectHelper';

const ProjectRepoListItem: React.FC<ProjectRepoListItemProps> = ({
  repo,
  activeProject,
  projectList,
  dispatch,
}) => {
  const [isChecked, setIsChecked] = useState(repo.isIncluded || false);

  useEffect(() => {
    setIsChecked(repo.isIncluded || false);
  });

  const toggleIsIncluded = (): void => {
    // define current repo
    const currentProject: Project = findActiveProject(
      projectList,
      activeProject
    );

    // need to make copy of states to include in new to not mutate state directly
    // make copy of repo with toggled isIncluded value
    const newRepo: Repo = { ...repo, isIncluded: !repo.isIncluded };

    // make copy of active project repo list with new repo included
    const newProjectRepos = currentProject.projectRepos.map((repo) =>
      repo.repoId === newRepo.repoId ? newRepo : repo
    );

    // copy active project
    const newActiveProject: Project = { ...currentProject };
    newActiveProject.projectRepos = newProjectRepos;

    // insert new project into new project list
    const newProjectList: Project[] = projectList.map((project) =>
      project.projectId === newActiveProject.projectId
        ? newActiveProject
        : project
    );

    // set new project list
    dispatch({ type: 'addRepo', payload: newProjectList });
  };

  return (
    <div onClick={toggleIsIncluded} style={{ margin: '10px' }}>
      {/* checkbox is readonly because it rerenders on state change */}
      <input type='checkbox' checked={isChecked} readOnly />
      {repo.repoName}
    </div>
  );
};

export default ProjectRepoListItem;
