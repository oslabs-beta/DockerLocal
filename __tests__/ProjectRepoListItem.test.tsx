import * as React from 'react';
import { shallow, mount } from 'enzyme';
import ProjectRepoListItem from '../src/client/components/projects/ProjectRepoListItem';

describe('<ProjectRepoListItem/> unit test', () => {

  const propsRepo = {
    repoName: 'mockRepoName',
    repoId: 'mockRepoId',
    repoOwner: 'mockRepoOwner',
    isIncluded: false
  }

  const propsProject = {
    projectName: 'mockPorjectName',
    projectId: 'mockPorjectId',
    projectRepos: [propsRepo]
  }

  const propsProjectRepoList = {
    activeProject: "repostring",
    repo: propsRepo,
    projectList: [propsProject],
    dispatch: jest.fn()
  }

  const wrapper = shallow(<ProjectRepoListItem {...propsProjectRepoList} />);

  it('should render a <div> and a <input>', () => {
    expect(wrapper.find('div')).toHaveLength(1);
    expect(wrapper.find('input')).toHaveLength(1);
  })

  it('should render a repo name', () => {
    expect(wrapper.find('div').text()).toEqual(propsRepo.repoName);

  })

  it('should matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  // TODO:
  xit('should possible to activate input checkbox', () => {
    // wrapper.find('div').simulate('click');
    // expect(propsRepo.isIncluded).toBe(true);
  })
})