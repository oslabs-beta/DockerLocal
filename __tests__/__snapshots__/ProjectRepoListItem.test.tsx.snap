// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectRepoListItem/> unit test should matches the snapshot 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "margin": "10px",
    }
  }
>
  <input
    checked={false}
    readOnly={true}
    type="checkbox"
  />
  mockRepoName
</div>
`;
